sway-core/src/type_engine/unresolved_type_check.rs

```diff
@@ -0,0 +1,248 @@
+//! This module handles the process of iterating through the typed AST and ensuring that all types
+//! are well-defined and well-formed. This process is run on the AST before we pass it into the IR,
+//! as the IR assumes all types are well-formed and will throw an ICE (internal compiler error) if
+//! that is not the case.
+
+use crate::{semantic_analysis::*, type_engine::*};
+
+/// If any types contained by this node are unresolved or have yet to be inferred, throw an
+/// error to signal to the user that more type information is needed.
+pub(crate) trait UnresolvedTypeCheck {
+    fn check_for_unresolved_types(&self) -> Vec<CompileError>;
+}
+
+impl UnresolvedTypeCheck for TypeId {
+    fn check_for_unresolved_types(&self) -> Vec<CompileError> {
+        use TypeInfo::*;
+        match look_up_type_id(*self) {
+            UnknownGeneric { name } => vec![CompileError::UnableToInferGeneric {
```

Comment:
Small nit regarding how this error is defined in `errors.rs`. Should we change it to be closer to Rust? Something like: 
"cannot infer type for type parameter `T`..."